### ISSUE TITLE
Consider new architecture to move all alerts to real time alerts

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#2679]
We're planning a big effort to bring performant real-time alerts to the RECAP Archive. This should be pretty powerful, and if we're going to do that, and if we're going to make it performant, we should just do it everywhere. The current plan is to do oral arguments, then people, then opinions, then RECAP, so when we overhaul oral arguments, we should move its alerts to real time.

One thing to note: Sometimes people like getting a batch of alerts at the end of the day/week/month. For those people, we should identify the hit by querying ES right when the data is ingested, but we should set it aside until the end of the day/ week/month when we send them one (and only one) email.

**Comment by mlissner** [comment]
Something else to think about is that maybe we want to have docket-level alerts and document-level alerts for RECAP. We have a client that wants to know about every new patent case. I'm not sure how they'd know that with the current flattened RECAP architecture. They could easily learn about every new RECAP *document*, but learning about new RECAP *cases* seems hard. 

I think this is tied up with the need to somehow do joining during our queries of the RECAP archive. If we can figure that out, I suspect we'd wind up with a docket index and a document index, which should make it easier to do docket- or document-based alerts.

**Comment by mlissner** [comment]
@albertisfu I'm sure you're thinking more about this and experimenting. If we wind up needing to have two indexes for this, like I said in Slack, it'd be great to see who used the RECAP search API, and maybe I can reach out to them and start talking to them to see what users think. It's interesting because the search API is already flat, so maybe this isn't so bad. It'd mean removing a number of fields from that API, and then creating a new one that has those fields.

**Comment by albertisfu** [comment]
Yeah, definitely it'd be good to ask users what they think about having two indexes for RECAP.
The alternative is to try to solve the problem of dockets without documents being searchable, there is a query called a `nested query` in ES that sounds like it can help, but I need to do some testing and confirm.

About RT search alerts, I've managed to make the percolator work in our Oral arguments implementation. It didn't require an additional index for storing the percolator queries, actually is required the percolator field is in the same document (OA) index.
I'm close to having the first version for RT alerts, I'm just dealing with the Django ES DSL `post_index` signal which is not working as expected.

Daily, Weekly, and monthly alerts, will require more work. If we want to use the percolator for these will need to store RT alerts to send them later according to alert frequency. I'll be posting the possible solutions, so we can decide since it might be simpler to just perform the search and return results for Daily, Weekly, and monthly alerts.

**Comment by mlissner** [comment]
@albertisfu, can you please share the status of making all indexes do percolator-based alerts? Do we need this issue anymore?

**Comment by albertisfu** [comment]
Sure. We have completed migrating all the Alerts code to Elasticsearch. Once #3632 is completed, all types of Search Alerts will work on ES.

However, Opinions Alerts will continue using the same search approach as in Solr. We have this issue, https://github.com/freelawproject/courtlistener/issues/4247, to move Opinions Alerts to the Percolator/Sweep index approach.

So I believe this issue can be closed and just follow up the Opinions one.

**Comment by mlissner** [comment]
Wonderful, that's sorta what I thought. Will close, thank you.
